Describe fetching plugin dependencies (#203)
kubeshop/botkube-docs: docs/plugin/dependencies.md

---
id: dependencies
title: "Dependencies"
sidebar_position: 5
---

Plugins can depend on other binaries, which are then downloaded by Botkube along with a given plugin. This is supported for both executor and source plugins.

> For example,

Reviewer: ? 😄

## Define dependencies for plugin index generation

As a part of the `Metadata` method, define the `Dependencies` property. The key is the name of the dependency, and the value is a structure with links to binaries for each platform.

:::info
For downloading plugins and theirs dependencies, Botkube uses `go-getter` library. It supports multiple URL formats (such as HTTP, Git repositories or S3), and is able to unpack archives and extract binaries from them. For more details, see the documentation on the [`go-getter`](https://github.com/hashicorp/go-getter) GitHub repository.
:::

```go

> // Metadata returns details about Helm plugin.

Reviewer: sth is wrong with padding here

                }
			},
		},
	}, nil
}
```

Such a definition will result in the following `dependencies` section in the plugin index:

```yaml
- name: helm
  # ...
  urls:
    - url: "..." # URL to darwin/amd64 plugin binary
      platform:
        os: darwin
        architecture: amd64
      dependencies:
        helm:
          url: https://get.helm.sh/helm-v3.6.3-darwin-amd64.tar.gz//darwin-amd64
    - url: "..." # URL to darwin/amd64 plugin binary
      platform:
        os: darwin
        architecture: arm64
      dependencies:
        helm:
          url: https://get.helm.sh/helm-v3.6.3-darwin-arm64.tar.gz//darwin-arm64
    - url: "..." # URL to linux/amd64 plugin binary
      platform:
        os: linux
        architecture: amd64
      dependencies:
        helm:
          url: https://get.helm.sh/helm-v3.6.3-linux-amd64.tar.gz//linux-amd64
    - url: "..." # URL to linux/arm64 plugin binary
      platform:
        os: linux
        architecture: arm64
      dependencies:
        helm:
          url: https://get.helm.sh/helm-v3.6.3-linux-arm64.tar.gz//linux-arm64
```

Read how to use the plugin index in the [Repository](./repository.md) document.

## Using dependencies

> During Botkube startup, Botkube plugin manager will fetch the plugin binaries along with all dependencies. Each dependency binary will be named exactly as specified in the [plugin index](#define-dependencies-for-plugin-index-generation). The dependency will be fetched to a directory specified in the `PLUGIN_DEPENDENCY_DIR` environment variable passed to the plugin.

Reviewer: ```suggestion
During Botkube startup, Botkube plugin manager fetches the plugin binaries along with all dependencies. Each dependency binary is named exactly as specified in the [plugin index](#define-dependencies-for-plugin-index-generation). The dependency is fetched to a directory specified in the `PLUGIN_DEPENDENCY_DIR` environment variable passed to the plugin.
```

IMO it's easier to read this way and more consistent with other parts of docs

For example, in case of Helm executor, to run the Helm binary fetched by Botkube plugin manager, the plugin uses the following code:

```go
	commandName := "helm"
	depDir, found := os.LookupEnv("PLUGIN_DEPENDENCY_DIR")

> 	if found {

Reviewer: It will be helpful to add some comments describing this approach
